File: cmd/nomi/plan_fmt_test.go
package main

import (
	"strings"
	"testing"
)

func TestFormatPlanReview_IncludesGoalStepsAndDiff(t *testing.T) {
	plan := &planPayload{Steps: []planStep{
		{
			Title: "Apply fix", ExpectedTool: "filesystem.patch", ExpectedCapability: "filesystem.write",
			Arguments: map[string]any{
				"diff": "--- a/foo.go\n+++ b/foo.go\n@@ -1 +1 @@\n-old\n+new\n",
			},
		},
		{
			Title: "Read README", ExpectedTool: "filesystem.read", ExpectedCapability: "filesystem.read",
			Arguments: map[string]any{"path": "README.md"},
		},
	}}
	var b strings.Builder
	formatPlanReview(&b, "Fix the flaky test", plan)
	out := b.String()
	if !strings.Contains(out, "Fix the flaky test") {
		t.Fatalf("goal missing: %s", out)
	}
	if !strings.Contains(out, "Apply fix") || !strings.Contains(out, "filesystem.write") {
		t.Fatalf("patch step missing: %s", out)
	}
	if !strings.Contains(out, "--- a/foo.go") {
		t.Fatalf("diff body missing: %s", out)
	}
	if !strings.Contains(out, "read: README.md") {
		t.Fatalf("read path missing: %s", out)
	}
	if !strings.Contains(out, "writes files") {
		t.Fatalf("caution banner missing: %s", out)
	}
}

func TestPlanRequiresCaution(t *testing.T) {
	safe := &planPayload{Steps: []planStep{{
		ExpectedTool: "filesystem.read", ExpectedCapability: "filesystem.read",
	}}}
	if planRequiresCaution(safe) {
		t.Fatal("read-only must not caution")
	}
	write := &planPayload{Steps: []planStep{{
		ExpectedTool: "filesystem.write", ExpectedCapability: "filesystem.write",
	}}}
	if !planRequiresCaution(write) {
		t.Fatal("write must caution")
	}
	rm := &planPayload{Steps: []planStep{{
		ExpectedTool: "command.exec", ExpectedCapability: "command.exec",
		Arguments: map[string]any{"command": "rm -rf /tmp/x"},
	}}}
	if !planRequiresCaution(rm) {
		t.Fatal("rm -rf must caution")
	}
}

func TestFormatPlanReview_CommandExec(t *testing.T) {
	plan := &planPayload{Steps: []planStep{{
		Title: "Run tests", ExpectedTool: "command.exec", ExpectedCapability: "command.exec",
		Arguments: map[string]any{"command": "go test ./..."},
	}}}
	var b strings.Builder
	formatPlanReview(&b, "test", plan)
	if !strings.Contains(b.String(), "$ go test ./...") {
		t.Fatalf("command missing: %s", b.String())
	}
}

File: extensions/vscode/src/plan_fmt.test.ts
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import type { Plan } from "./client.js";
import {
  formatPlanReview,
  planRequiresCaution,
  summarizeDiff,
} from "./plan_fmt.js";

describe("summarizeDiff", () => {
  it("counts lines and files", () => {
    const diff = `--- a/foo.ts
+++ b/foo.ts
@@ -1,2 +1,3 @@
 keep
-old
+new
+more
`;
    const s = summarizeDiff(diff);
    assert.deepEqual(s.files, ["foo.ts"]);
    assert.equal(s.added, 2);
    assert.equal(s.removed, 1);
  });
});

describe("planRequiresCaution", () => {
  it("flags write and patch", () => {
    const plan: Plan = {
      id: "p1",
      version: 1,
      steps: [
        {
          id: "s1",
          title: "patch",
          expected_tool: "filesystem.patch",
          order: 0,
          arguments: { diff: "--- a/x\n+++ b/x\n+hi\n" },
        },
      ],
    };
    assert.equal(planRequiresCaution(plan), true);
  });

  it("ignores read-only", () => {
    const plan: Plan = {
      id: "p1",
      version: 1,
      steps: [
        {
          id: "s1",
          title: "read",
          expected_tool: "filesystem.read",
          expected_capability: "filesystem.read",
          order: 0,
          arguments: { path: "/tmp/a" },
        },
      ],
    };
    assert.equal(planRequiresCaution(plan), false);
  });
});

describe("formatPlanReview", () => {
  it("includes goal, steps, and diff body", () => {
    const plan: Plan = {
      id: "p1",
      version: 1,
      steps: [
        {
          id: "s1",
          title: "Apply fix",
          expected_tool: "filesystem.patch",
          expected_capability: "filesystem.write",
          why: "user asked",
          order: 0,
          arguments: {
            diff: "--- a/a.ts\n+++ b/a.ts\n@@ -1 +1 @@\n-old\n+new\n",
          },
        },
      ],
    };
    const text = formatPlanReview("fix the bug", plan);
    assert.match(text, /Goal: fix the bug/);
    assert.match(text, /Apply fix/);
    assert.match(text, /diff: \+1 −1/);
    assert.match(text, /\+new/);
    assert.match(text, /writes files/);
  });
});
